File: app/metrics/deadlines.py
"""Tax deadlines of a freelancer liable to VAT (Germany), from space settings.

    settings["tax"] = {"vat": {"return_interval": "monthly", "extension": true},
                       "prepayments": {"amount": 1200}, "annual_due": "07-31"}
"""

from datetime import date, timedelta

from app.metrics.dates import add_months

DUE_DAY = 10
PREPAYMENT_MONTHS = (3, 6, 9, 12)
DEFAULT_ANNUAL = "07-31"
QUARTER = 3


def _vat_returns(today: date, tax: dict, horizon: date) -> list[dict]:
    """Voranmeldung due on the 10th after the period (+1 month with Dauerfristverlängerung)."""
    vat = tax.get("vat") or {}
    months = QUARTER if vat.get("return_interval") == "quarterly" else 1
    shift = 1 if vat.get("extension") else 0
    found = []
    start = add_months(today, -12)
    while start <= horizon:
        if months == 1 or (start.month - 1) % QUARTER == 0:
            period_end = add_months(start, months) - timedelta(days=1)
            due = add_months(period_end, 1 + shift).replace(day=DUE_DAY)
            if today <= due <= horizon:
                label = start.strftime("%m/%Y") if months == 1 else f"Q{(start.month - 1) // 3 + 1}/{start.year}"
                found.append({"kind": "vat_return", "due": due, "period": label,
                              "period_start": start, "period_end": period_end})
        start = add_months(start, 1)
    return found


def _prepayments(today: date, tax: dict, horizon: date) -> list[dict]:
    amount = (tax.get("prepayments") or {}).get("amount")
    found = []
    for year in (today.year, today.year + 1):
        for month in PREPAYMENT_MONTHS:
            due = date(year, month, DUE_DAY)
            if today <= due <= horizon:
                found.append({"kind": "prepayment", "due": due, "amount": amount})
    return found


def _annual(today: date, tax: dict, horizon: date) -> list[dict]:
    """VAT and income tax returns for last year (default: 31 July, without tax advisor)."""
    month, day_ = (int(x) for x in str(tax.get("annual_due") or DEFAULT_ANNUAL).split("-"))
    found = []
    for year in (today.year, today.year + 1):
        due = date(year, month, day_)
        if today <= due <= horizon:
            found.append({"kind": "annual", "due": due, "year": year - 1})
    return found


def upcoming(settings: dict, today: date, days: int) -> list[dict]:
    tax = settings.get("tax") or {}
    horizon = today + timedelta(days=days)
    items = _vat_returns(today, tax, horizon) + _prepayments(today, tax, horizon) + _annual(today, tax, horizon)
    return sorted(items, key=lambda i: i["due"])

File: goapp/internal/metrics/deadlines.go
// Package metrics: tax deadlines of a freelancer liable to VAT (Germany),
// from space settings.
//
//	settings["tax"] = {"vat": {"return_interval": "monthly", "extension": true},
//	                   "prepayments": {"amount": 1200}, "annual_due": "07-31"}
package metrics

import (
	"fmt"
	"time"
)

const (
	dueDay        = 10
	defaultAnnual = "07-31"
	quarterMonths = 3
)

var prepaymentMonths = []time.Month{time.March, time.June, time.September, time.December}

// TaxSettings mirrors the "tax" key of a space's settings map.
type TaxSettings struct {
	VATReturnInterval string // "monthly" (default) or "quarterly"
	VATExtension      bool   // Dauerfristverlängerung: due date shifts by one month
	PrepaymentAmount  *float64
	AnnualDueMonthDay string // "MM-DD", default "07-31"
}

// TaxDeadline is one upcoming deadline.
type TaxDeadline struct {
	Kind        string // "vat_return", "prepayment", "annual"
	Due         time.Time
	Period      string    // vat_return only: "09/2026" or "Q3/2026"
	PeriodStart time.Time // vat_return only
	PeriodEnd   time.Time // vat_return only
	Amount      *float64  // prepayment only
	Year        int       // annual only: the year the return covers
}

func vatReturns(today time.Time, tax TaxSettings, horizon time.Time) []TaxDeadline {
	months := 1
	if tax.VATReturnInterval == "quarterly" {
		months = quarterMonths
	}
	shift := 0
	if tax.VATExtension {
		shift = 1
	}

	var found []TaxDeadline
	for start := AddMonths(today, -12); !start.After(horizon); start = AddMonths(start, 1) {
		if months != 1 && (int(start.Month())-1)%quarterMonths != 0 {
			continue
		}
		periodEnd := AddMonths(start, months).AddDate(0, 0, -1)
		due := time.Date(AddMonths(periodEnd, 1+shift).Year(), AddMonths(periodEnd, 1+shift).Month(), dueDay,
			0, 0, 0, 0, time.UTC)
		if today.After(due) || due.After(horizon) {
			continue
		}
		label := start.Format("01/2006")
		if months != 1 {
			label = fmt.Sprintf("Q%d/%d", (int(start.Month())-1)/3+1, start.Year())
		}
		found = append(found, TaxDeadline{
			Kind: "vat_return", Due: due, Period: label, PeriodStart: start, PeriodEnd: periodEnd,
		})
	}
	return found
}

func prepayments(today time.Time, tax TaxSettings, horizon time.Time) []TaxDeadline {
	var found []TaxDeadline
	for _, year := range []int{today.Year(), today.Year() + 1} {
		for _, month := range prepaymentMonths {
			due := time.Date(year, month, dueDay, 0, 0, 0, 0, time.UTC)
			if !today.After(due) && !due.After(horizon) {
				found = append(found, TaxDeadline{Kind: "prepayment", Due: due, Amount: tax.PrepaymentAmount})
			}
		}
	}
	return found
}

func annual(today time.Time, tax TaxSettings, horizon time.Time) []TaxDeadline {
	spec := tax.AnnualDueMonthDay
	if spec == "" {
		spec = defaultAnnual
	}
	var month, day int
	fmt.Sscanf(spec, "%d-%d", &month, &day)
	if month == 0 {
		month, day = 7, 31
	}

	var found []TaxDeadline
	for _, year := range []int{today.Year(), today.Year() + 1} {
		due := time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
		if !today.After(due) && !due.After(horizon) {
			found = append(found, TaxDeadline{Kind: "annual", Due: due, Year: year - 1})
		}
	}
	return found
}

// UpcomingDeadlines returns every tax deadline within `days`, soonest first.
func UpcomingDeadlines(tax TaxSettings, today time.Time, days int) []TaxDeadline {
	horizon := today.AddDate(0, 0, days)
	items := append(vatReturns(today, tax, horizon), prepayments(today, tax, horizon)...)
	items = append(items, annual(today, tax, horizon)...)

	for i := 1; i < len(items); i++ {
		for j := i; j > 0 && items[j].Due.Before(items[j-1].Due); j-- {
			items[j], items[j-1] = items[j-1], items[j]
		}
	}
	return items
}
